File: android/library/src/main/kotlin/ge/silknet/ds/components/SilknetIconButton.kt
// SilknetIconButton — square, single-icon button. Mirrors iOS SilknetIconButton.
// Verified against Figma node 2052:3517. Uniform padding = (size − icon) / 2.

package ge.silknet.ds.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.LocalContentColor
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import ge.silknet.ds.Silknet
import ge.silknet.ds.SilknetTheme

enum class SilknetIconButtonVariant { Primary, PrimarySoft, Secondary, Ghost }

@Composable
fun SilknetIconButton(
    contentDescription: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    variant: SilknetIconButtonVariant = SilknetIconButtonVariant.Primary,
    size: SilknetButtonSize = SilknetButtonSize.Md,
    enabled: Boolean = true,
    icon: @Composable () -> Unit,
) {
    val interactionSource = remember { MutableInteractionSource() }
    val isPressed by interactionSource.collectIsPressedAsState()

    val bg = iconBgFor(variant, enabled)
    val fg = iconFgFor(variant, enabled)
    val borderColor = iconBorderFor(variant, enabled)
    val shape = RoundedCornerShape(iconCornerFor(size))
    val iconSizeDp = iconSizeFor(size)
    val pad = iconPadFor(size)

    Box(
        modifier = modifier
            .semantics { this.contentDescription = contentDescription }
            .clip(shape)
            .background(bg)
            .let { if (borderColor != null) it.border(1.dp, borderColor, shape) else it }
            .clickable(
                interactionSource = interactionSource,
                indication = null,
                enabled = enabled,
                onClick = onClick,
            )
            .padding(pad),
        contentAlignment = Alignment.Center,
    ) {
        if (isPressed && enabled) {
            Box(
                modifier = Modifier
                    .matchParentSize()
                    .background(Color.Black.copy(alpha = 0.16f))
            )
        }
        Box(modifier = Modifier.size(iconSizeDp)) {
            CompositionLocalProvider(LocalContentColor provides fg) { icon() }
        }
    }
}

@Composable
private fun iconBgFor(variant: SilknetIconButtonVariant, enabled: Boolean): Color {
    if (!enabled) {
        return when (variant) {
            SilknetIconButtonVariant.Primary, SilknetIconButtonVariant.PrimarySoft -> Silknet.colors.backgroundDisabled
            SilknetIconButtonVariant.Secondary, SilknetIconButtonVariant.Ghost -> Color.Transparent
        }
    }
    return when (variant) {
        SilknetIconButtonVariant.Primary -> Silknet.colors.backgroundPrimaryAccent
        SilknetIconButtonVariant.PrimarySoft -> Silknet.colors.backgroundPrimarySoft
        SilknetIconButtonVariant.Secondary -> Silknet.colors.backgroundLayer
        SilknetIconButtonVariant.Ghost -> Color.Transparent
    }
}

@Composable
private fun iconFgFor(variant: SilknetIconButtonVariant, enabled: Boolean): Color {
    if (!enabled) {
        return when (variant) {
            SilknetIconButtonVariant.Primary, SilknetIconButtonVariant.PrimarySoft -> Silknet.colors.textContrast
            SilknetIconButtonVariant.Secondary, SilknetIconButtonVariant.Ghost -> Silknet.colors.textDisabled
        }
    }
    return when (variant) {
        SilknetIconButtonVariant.Primary -> Silknet.colors.textContrast
        SilknetIconButtonVariant.PrimarySoft -> Silknet.colors.textPrimary
        SilknetIconButtonVariant.Secondary, SilknetIconButtonVariant.Ghost -> Silknet.colors.textSecondary
    }
}

@Composable
private fun iconBorderFor(variant: SilknetIconButtonVariant, enabled: Boolean): Color? {
    if (variant != SilknetIconButtonVariant.Secondary) return null
    return if (enabled) Silknet.colors.borderDefault else Silknet.colors.backgroundDisabled
}

@Composable
private fun iconCornerFor(size: SilknetButtonSize): Dp = when (size) {
    SilknetButtonSize.Xs, SilknetButtonSize.Sm -> Silknet.dimens.digitsRadiusXs
    SilknetButtonSize.Md, SilknetButtonSize.Lg -> Silknet.dimens.digitsRadiusS
}

private fun iconSizeFor(size: SilknetButtonSize): Dp = when (size) {
    SilknetButtonSize.Xs, SilknetButtonSize.Sm -> 16.dp
    SilknetButtonSize.Md, SilknetButtonSize.Lg -> 20.dp
}

@Composable
private fun iconPadFor(size: SilknetButtonSize): Dp = when (size) {
    SilknetButtonSize.Xs -> Silknet.dimens.digitsSpacing1
    SilknetButtonSize.Sm -> Silknet.dimens.digitsSpacing2
    SilknetButtonSize.Md -> Silknet.dimens.digitsSpacing3
    SilknetButtonSize.Lg -> Silknet.dimens.digitsSpacing4
}

@Preview(name = "All variants — md")
@Composable
private fun PreviewIconAllVariants() {
    SilknetTheme {
        Box(
            modifier = Modifier
                .background(Silknet.colors.backgroundLayer)
                .padding(24.dp)
        ) {
            Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                SilknetIconButton("Add", {}, variant = SilknetIconButtonVariant.Primary) { PlusGlyph() }
                SilknetIconButton("Add", {}, variant = SilknetIconButtonVariant.PrimarySoft) { PlusGlyph() }
                SilknetIconButton("Add", {}, variant = SilknetIconButtonVariant.Secondary) { PlusGlyph() }
                SilknetIconButton("Add", {}, variant = SilknetIconButtonVariant.Ghost) { PlusGlyph() }
            }
        }
    }
}

@Preview(name = "Sizes")
@Composable
private fun PreviewIconSizes() {
    SilknetTheme {
        Box(
            modifier = Modifier
                .background(Silknet.colors.backgroundLayer)
                .padding(24.dp)
        ) {
            Row(horizontalArrangement = Arrangement.spacedBy(12.dp), verticalAlignment = Alignment.CenterVertically) {
                SilknetIconButton("xs", {}, size = SilknetButtonSize.Xs) { PlusGlyph() }
                SilknetIconButton("sm", {}, size = SilknetButtonSize.Sm) { PlusGlyph() }
                SilknetIconButton("md", {}, size = SilknetButtonSize.Md) { PlusGlyph() }
                SilknetIconButton("lg", {}, size = SilknetButtonSize.Lg) { PlusGlyph() }
            }
        }
    }
}

// Tiny stand-in icon so previews don't depend on an icons dependency.
@Composable
private fun PlusGlyph() {
    Box(
        modifier = Modifier
            .size(20.dp),
        contentAlignment = Alignment.Center,
    ) {
        Box(Modifier.size(width = 12.dp, height = 2.dp).background(LocalContentColor.current))
        Box(Modifier.size(width = 2.dp, height = 12.dp).background(LocalContentColor.current))
    }
}

File: Sources/SilknetDS/Components/SilknetIconButton.swift
// SilknetIconButton — square Button with a single icon, no text.
// Verified against Figma node 2052:3517. Reuses the SilknetButton color
// system but with uniform padding and icon-only content.
//
//     SilknetIconButton(label: "Add", variant: .primary) {
//         Image(systemName: "plus")
//     } action: { /* onTap */ }

import SwiftUI

public enum SilknetIconButtonVariant: Sendable {
    case primary
    case primarySoft
    case secondary
    case ghost
}

public struct SilknetIconButton<Icon: View>: View {
    private let label: String
    private let variant: SilknetIconButtonVariant
    private let size: SilknetButtonSize
    private let icon: () -> Icon
    private let action: () -> Void

    public init(
        label: String,
        variant: SilknetIconButtonVariant = .primary,
        size: SilknetButtonSize = .md,
        @ViewBuilder icon: @escaping () -> Icon,
        action: @escaping () -> Void = {}
    ) {
        self.label = label
        self.variant = variant
        self.size = size
        self.icon = icon
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            icon()
                .frame(width: iconSize, height: iconSize)
        }
        .buttonStyle(SilknetIconButtonStyle(variant: variant, size: size))
        .accessibilityLabel(label)
    }

    private var iconSize: CGFloat {
        switch size {
        case .xs, .sm: return 16
        case .md, .lg: return 20
        }
    }
}

private struct SilknetIconButtonStyle: ButtonStyle {
    let variant: SilknetIconButtonVariant
    let size: SilknetButtonSize

    @Environment(\.isEnabled) private var isEnabled

    func makeBody(configuration: Configuration) -> some View {
        let pressed = configuration.isPressed
        let shape = RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)

        return configuration.label
            .padding(uniformPadding)
            .foregroundColor(foregroundColor)
            .background(
                ZStack {
                    backgroundColor.clipShape(shape)
                    if pressed && isEnabled {
                        Color.black.opacity(0.16).clipShape(shape)
                    }
                }
            )
            .overlay(borderOverlay(shape: shape))
            .contentShape(shape)
    }

    private var uniformPadding: CGFloat {
        // Per Figma: icon sits centered with padding = (size − icon) / 2.
        // xs(24): icon 16, pad 4. sm(32): icon 16, pad 8. md(44): icon 20, pad 12. lg(52): icon 20, pad 16.
        switch size {
        case .xs: return .silknet.digitsSpacing1
        case .sm: return .silknet.digitsSpacing2
        case .md: return .silknet.digitsSpacing3
        case .lg: return .silknet.digitsSpacing4
        }
    }

    private var cornerRadius: CGFloat {
        switch size {
        case .xs, .sm: return .silknet.digitsRadiusXs
        case .md, .lg: return .silknet.digitsRadiusS
        }
    }

    private var backgroundColor: Color {
        if !isEnabled {
            switch variant {
            case .secondary, .ghost: return .clear
            case .primary, .primarySoft: return .silknet.backgroundDisabled
            }
        }
        switch variant {
        case .primary:     return .silknet.backgroundPrimaryAccent
        case .primarySoft: return .silknet.backgroundPrimarySoft
        case .secondary:   return .silknet.backgroundLayer
        case .ghost:       return .clear
        }
    }

    private var foregroundColor: Color {
        if !isEnabled {
            switch variant {
            case .primary, .primarySoft: return .silknet.textContrast
            case .secondary, .ghost:     return .silknet.textDisabled
            }
        }
        switch variant {
        case .primary:           return .silknet.textContrast
        case .primarySoft:       return .silknet.textPrimary
        case .secondary, .ghost: return .silknet.textSecondary
        }
    }

    @ViewBuilder
    private func borderOverlay(shape: RoundedRectangle) -> some View {
        if !isEnabled, variant == .secondary {
            shape.strokeBorder(Color.silknet.backgroundDisabled, lineWidth: 1)
        } else if isEnabled, variant == .secondary {
            shape.strokeBorder(Color.silknet.borderDefault, lineWidth: 1)
        }
    }
}

#if DEBUG
#Preview("All variants — md") {
    HStack(spacing: 16) {
        SilknetIconButton(label: "Add", variant: .primary)     { Image(systemName: "plus") }
        SilknetIconButton(label: "Add", variant: .primarySoft) { Image(systemName: "plus") }
        SilknetIconButton(label: "Add", variant: .secondary)   { Image(systemName: "plus") }
        SilknetIconButton(label: "Add", variant: .ghost)       { Image(systemName: "plus") }
    }
    .padding(24)
    .background(Color.silknet.backgroundLayer)
}

#Preview("Sizes — primary") {
    HStack(spacing: 16) {
        SilknetIconButton(label: "xs", size: .xs) { Image(systemName: "plus") }
        SilknetIconButton(label: "sm", size: .sm) { Image(systemName: "plus") }
        SilknetIconButton(label: "md", size: .md) { Image(systemName: "plus") }
        SilknetIconButton(label: "lg", size: .lg) { Image(systemName: "plus") }
    }
    .padding(24)
    .background(Color.silknet.backgroundLayer)
}

#Preview("Disabled") {
    HStack(spacing: 16) {
        SilknetIconButton(label: "Add", variant: .primary)     { Image(systemName: "plus") }.disabled(true)
        SilknetIconButton(label: "Add", variant: .primarySoft) { Image(systemName: "plus") }.disabled(true)
        SilknetIconButton(label: "Add", variant: .secondary)   { Image(systemName: "plus") }.disabled(true)
        SilknetIconButton(label: "Add", variant: .ghost)       { Image(systemName: "plus") }.disabled(true)
    }
    .padding(24)
    .background(Color.silknet.backgroundLayer)
}
#endif
